pre_commit/store.py

```diff
@@ -134,19 +132,50 @@ def _get_result():
                 )
         return directory
 
-    def clone(self, repo, ref, deps=()):
-        """Clone the given url and checkout the specific ref."""
-        def clone_strategy(directory):
-            env = git.no_git_env()
+    def _perform_safe_clone(self, clone_strategy):
+        directory = tempfile.mkdtemp(prefix='repo', dir=self.directory)
+        with clean_path_on_failure(directory):
+            clone_strategy(directory)
+        return directory
 
-            cmd = ('git', 'clone', '--no-checkout', repo, directory)
-            cmd_output(*cmd, env=env)
+    def _complete_clone(self, repo, ref, directory):
+        """Perform a complete clone of a repository and its submodules """
+        env = git.no_git_env()
 
-            def _git_cmd(*args):
-                return cmd_output('git', *args, cwd=directory, env=env)
+        cmd = ('git', 'clone', '--no-checkout', repo, directory)
+        cmd_output(*cmd, env=env)
+
+        def _git_cmd(*args):
+            return cmd_output('git', *args, cwd=directory, env=env)
+
+        _git_cmd('reset', ref, '--hard')
+        _git_cmd('submodule', 'update', '--init', '--recursive')
 
-            _git_cmd('reset', ref, '--hard')
-            _git_cmd('submodule', 'update', '--init', '--recursive')
+    def _shallow_clone(self, repo, ref, directory):
+        """Perform a shallow clone of a repository and its submodules """
+        env = git.no_git_env()
+
+        def _git_cmd(*args):
+            return cmd_output('git', *args, cwd=directory, env=env)
+
+        _git_cmd('init', '.')
+        _git_cmd('remote', 'add', 'origin', repo)
+        _git_cmd('fetch', 'origin', ref, '--depth=1')
```

Comment:
I think there's actually three strategies we have to try here:

1. fetch --depth=1
2. set the git protocol to v2 and then fetch --depth=1 (? or maybe just try this first since it's the most reliable for github)
3. fall back to full clone

Reply:
I'll add `git -c protocol.version=2` to the `_git_cmd` for the second option.